File: R/forLater/extract.R
#' Extract a sub-sample of data from a vector or matrix.
#' 
#' @param x A vector or matrix of measurements. If x is a matrix, each column is treated as a separate measurement vector.
#' @param fs the sampling rate in Hz of the data in x.
#' @param tst Defines the start time in seconds of the interval to be extracted from x.
#' @param ted Defines the end time in seconds of the interval to be extracted from x.
#' @return X A matrix containing a sub-sample of x. X has the same number of columns as x. The length of the sub-sample will be round(fs*(tend-tstart)) samples.
#' @note Output sampling rate is the same as the input sampling rate.
#' @note If either tst or ted are beyond the length of x, non-existing samples will be replaced with NaN in X.

extract <- function(x, fs, tst, ted) {
  X <- vector(mode="numeric", length=0)
  if (missing(ted)) {
    stop("inputs for all arguments are required")
  }
  if (nrow(x) == 1) {
    x <- t(x)
  }
  npre <- NULL
  npst <- NULL
  kst <- round(fs * tst) + 1
  ked <- round(fs * ted)
  if (kst > nrow(x)) {
    return(X)
  } else {
    if (kst < 0) {
      npre <- -kst
      kst <- 1
    }
    if (ked > nrow(x)) {
      npst <- ked - nrow(x)
      ked <- nrow(x)
    }
    if (!is.null(npre)) {
      X1 <- matrix(NaN, npre, ncol(x))
      X2 <- x[kst:ked, ]
      X3 <- matrix(NaN, npre, ncol(x))
      X <- rbind(X1, X2, X3)
    } else {
      X <- x[kst:ked, ]
    }
    return(X)
  }
}

File: matlab/not_for_distribution/new7July/processing/extract.m
function  X = extract(x,fs,tst,ted)

%     X = extract(x,fs,tst,ted)
%     Extract a sub-sample of data from a vector or matrix. 
%
%		Inputs:
%     x is a vector or matrix of measurements. If x is a matrix, each column
%		 is treated as a separate measurement vector.
%     fs is the sampling rate in Hz of the data in x.
%     tst defines the start time in seconds of the interval to be extracted from x.
%     ted defines the end time in seconds of the interval to be extracted from x.
%
%     Returns:
%     X is a matrix containing a sub-sample of x. X has the same number of columns 
%		 as x. The length of the sub-sample will be round(fs*(tend-tstart)) samples. 
%
%		Output sampling rate is the same as the input sampling rate.
%		If either tstart or tend are beyond the length of x, non-existing samples will 
%		be replaced with NaN in X. 
%
%		Example:
%		 X = extract(x,fs,tintvl)
% 	    returns: X=.
%
%     Valid: Matlab, Octave
%     markjohnson@st-andrews.ac.uk
%     Last modified: 10 May 2017

X = [] ;
if nargin<4,
   help extract
   return
end

if size(x,1)==1,
	x = x(:) ;
end
	
npre = [] ; npst = [] ;
kst = round(fs*tst)+1 ;
ked = round(fs*ted) ;
if kst>size(x,1),
	return
end
	
if kst<0,
	npre = -kst ;
	kst = 1 ;
end

if ked>size(x,1),
	npst = ked-size(x,1) ;
	ked = size(x,1) ;
end

X = [NaN*zeros(npre,size(X,2));x(kst:ked,:);NaN*zeros(npst,size(X,2))] ;
